src/store.ts

```diff
@@ -36,13 +36,29 @@ export class StateStore<T extends Record<string, unknown>> {
     };
   };
 
-  public subscribeWithSelector = <O extends readonly unknown[]>(selector: (nextValue: T) => O, handler: Handler<O>) => {
+  public subscribeWithSelector = <O extends Readonly<Record<string, unknown>> & { length?: never }>(
+    selector: (nextValue: T) => O,
+    handler: Handler<O>,
+  ) => {
     // begin with undefined to reduce amount of selector calls
     let selectedValues: O | undefined;
 
     const wrappedHandler: Handler<T> = (nextValue) => {
       const newlySelectedValues = selector(nextValue);
-      const hasUpdatedValues = selectedValues?.some((value, index) => value !== newlySelectedValues[index]) ?? true;
+
+      let hasUpdatedValues = !selectedValues;
+
+      if (Array.isArray(selectedValues)) {
+        console.warn(
+          'The API of our state store has changed. Instead of returning an array in the selector, please return a named object of properties.',
+        );
+      }
+
+      for (const key in selectedValues) {
+        if (selectedValues[key] === newlySelectedValues[key]) continue;
+        hasUpdatedValues = true;
+        break;
+      }
```

Comment:
Do we need this optimization? AFAIK, this comparison is broken, as the order of keys in the object isn't guaranteed.
This code will say that `{ a: 10, b: 20 }` is not equal to ` { b: 20, a: 10 }`.

@arnautov-anton, @isekovanic

Reply:
Hmm, but the order doesn't matter right ? We're just really looking for the keys and then extracting them from both places. We're just comparing the values for the specific key - we're not saying that `selectedValues[selectedValueKey]` needs to be equal to `newlySelectedValues[newlySelectedValuesKey]`.

Reply:
The only time this might fail would be if `selectedValues` and `newlySelectedValues` have different keys, which would never be the case because they're handled entirely by how the `selector` is defined. Or am I missing something ?

Reply:
Ignore me! You are completely right. Maybe loop over `newlySelectedValues` to make sure there aren't keys added or removed.

Reply:
Correct, the store is large and can hold multitude of other values that are selected but did not change - each selection produces new object and that object is being passed to `setState` - to prevent running `setState` each time unrelated change happens (value that changed has not been selected), we bail by comparing actually selected values.

Reply:
But, if you guys want I can also add something like `Object.keys(selectedValues).length === Object.keys(newlySelectedValues).length` just to be safe. We didn't have that check before either and I agree with @arnautov-anton that it would never happen, but maybe we should be safe.

Reply:
That check should be enough because if their sizes differ then they're by default different and an update should be triggered. If they're the same, only then do we check if all of the values are the same for the respective keys (if some keys differ it would be caught here).

Reply:
Removing or adding keys is obviously not recommended - it'll just trigger a re-render as either previous or newest wouldn't have the keys thus `if (selectedValues[key] === newlySelectedValues[key]) continue;` is no longer true.

Reply:
We can test just to be sure but I don't think adding an extra check is necessary.

Reply:
Yep agreed. I don't think we need to do it.